File: apps/roi-studio/roi_studio/core/exporters.py
"""Interop exports: COCO, YOLO, Pascal VOC and binary masks.

Every exporter takes canonical rows and writes into a subfolder of the batch,
so an export can never overwrite the batch's own outputs.  All of them go
through atomic_write and report rather than raise.
"""

from __future__ import annotations

import json
import os
from datetime import datetime
from xml.sax.saxutils import escape

from ..config import APP_VERSION
from . import geometry as geo
from .io_safe import WriteReport, atomic_write, ensure_dir, write_text_atomic

try:
    from PIL import Image, ImageDraw
    HAS_PIL = True
except Exception:                                    # pragma: no cover
    Image = ImageDraw = None
    HAS_PIL = False

CATEGORY_ID = 1
CATEGORY_NAME = "roi"


# ══════════════════════════════════════════════════════════════
def _roi_rows(rows):
    """Only rows that actually carry geometry."""
    out = []
    for row in rows:
        if row.get("row_type") != "roi":
            continue
        polys = geo.parse_multi_polys(row.get("pixel_coords", ""))
        if not polys:
            norm = geo.parse_multi_polys(row.get("normalized_coords", ""))
            w = int(row.get("image_width") or 0)
            h = int(row.get("image_height") or 0)
            if norm and w and h:
                polys = geo.norm_to_polys(norm, w, h)
        if polys:
            out.append((row, polys))
    return out


def _size_of(row, polys):
    """Image size from the row, falling back to the shapes' own extent."""
    w = int(row.get("image_width") or 0)
    h = int(row.get("image_height") or 0)
    if w and h:
        return w, h
    max_x = max((p[0] for poly in polys for p in poly), default=0)
    max_y = max((p[1] for poly in polys for p in poly), default=0)
    return int(max_x) + 1, int(max_y) + 1


def _bbox(poly):
    x0, y0, x1, y1 = geo.polygon_bounds(poly)
    return [float(x0), float(y0), float(x1 - x0), float(y1 - y0)]


# ══════════════════════════════════════════════════════════════
# COCO
# ══════════════════════════════════════════════════════════════
def export_coco(rows, folder, subdir="export_coco",
                filename="annotations.json") -> WriteReport:
    """One COCO instance-segmentation file for the whole batch."""
    report = WriteReport()
    out_dir = os.path.join(str(folder), subdir)
    if not ensure_dir(out_dir):
        report.errors.append("could not create %s" % subdir)
        return report

    images, annotations = [], []
    ann_id = 1
    for image_id, (row, polys) in enumerate(_roi_rows(rows), start=1):
        width, height = _size_of(row, polys)
        images.append({
            "id": image_id,
            "file_name": row.get("image_name", ""),
            "width": width,
            "height": height,
            "roi_key": row.get("roi_key", ""),
        })
        kinds = geo.parse_shape_types(row.get("shape_types", ""), len(polys))
        for i, poly in enumerate(polys):
            flat = [float(v) for pt in poly for v in pt]
            annotations.append({
                "id": ann_id,
                "image_id": image_id,
                "category_id": CATEGORY_ID,
                "segmentation": [flat],
                "bbox": _bbox(poly),
                "area": float(geo.polygon_area(poly)),
                "iscrowd": 0,
                "shape_type": kinds[i] if i < len(kinds) else "polygon",
            })
            ann_id += 1

    payload = {
        "info": {"description": "ROI Studio export",
                 "version": APP_VERSION,
                 "date_created": datetime.now().isoformat(timespec="seconds")},
        "licenses": [],
        "images": images,
        "annotations": annotations,
        "categories": [{"id": CATEGORY_ID, "name": CATEGORY_NAME,
                        "supercategory": "region"}],
    }
    path = os.path.join(out_dir, filename)
    ok, err = write_text_atomic(path, json.dumps(payload, indent=1),
                                verify_json=True, keep_backup=False)
    if ok:
        report.written.append(path)
        if not annotations:
            report.warnings.append("no ROI rows to export")
    else:
        report.errors.append("coco: %s" % err)
    return report


# ══════════════════════════════════════════════════════════════
# YOLO  (segmentation format)
# ══════════════════════════════════════════════════════════════
def export_yolo(rows, folder, subdir="export_yolo") -> WriteReport:
    """One .txt of normalised polygons per image, plus a dataset stub."""
    report = WriteReport()
    out_dir = os.path.join(str(folder), subdir)
    labels_dir = os.path.join(out_dir, "labels")
    if not ensure_dir(labels_dir):
        report.errors.append("could not create %s" % subdir)
        return report

    written = 0
    for row, polys in _roi_rows(rows):
        width, height = _size_of(row, polys)
        norm = geo.polys_to_norm(polys, width, height, ndigits=6)
        lines = []
        for poly in norm:
            if len(poly) < 3:
                continue
            coords = " ".join("%.6f %.6f" % (geo.clamp(x, 0.0, 1.0),
                                             geo.clamp(y, 0.0, 1.0))
                              for x, y in poly)
            lines.append("0 " + coords)
        if not lines:
            continue
        stem = os.path.splitext(row.get("image_name", ""))[0] or "image"
        path = os.path.join(labels_dir, stem + ".txt")
        ok, err = write_text_atomic(path, "\n".join(lines) + "\n",
                                    keep_backup=False)
        if ok:
            written += 1
        else:
            report.errors.append("%s: %s" % (os.path.basename(path), err))

    stub = os.path.join(out_dir, "data.yaml")
    ok, err = write_text_atomic(
        stub,
        "# ROI Studio export\n"
        "path: .\ntrain: images\nval: images\n\nnames:\n  0: %s\n" % CATEGORY_NAME,
        keep_backup=False)
    if ok:
        report.written.append(stub)
    report.written.append(labels_dir)
    if not written:
        report.warnings.append("no ROI rows to export")
    return report


# ══════════════════════════════════════════════════════════════
# PASCAL VOC
# ══════════════════════════════════════════════════════════════
def export_voc(rows, folder, subdir="export_voc") -> WriteReport:
    """One XML per image.

    VOC describes axis-aligned boxes, so each polygon is written as its
    bounding box and the exact points are kept in a custom <polygon> child
    for anything that can use them."""
    report = WriteReport()
    out_dir = os.path.join(str(folder), subdir)
    if not ensure_dir(out_dir):
        report.errors.append("could not create %s" % subdir)
        return report

    written = 0
    for row, polys in _roi_rows(rows):
        width, height = _size_of(row, polys)
        name = row.get("image_name", "")
        parts = ["<annotation>",
                 "  <folder>%s</folder>" % escape(os.path.basename(str(folder))),
                 "  <filename>%s</filename>" % escape(name),
                 "  <source><database>ROI Studio</database></source>",
                 "  <size><width>%d</width><height>%d</height>"
                 "<depth>3</depth></size>" % (width, height),
                 "  <segmented>1</segmented>"]
        kinds = geo.parse_shape_types(row.get("shape_types", ""), len(polys))
        for i, poly in enumerate(polys):
            x0, y0, x1, y1 = geo.polygon_bounds(poly)
            pts = " ".join("%d,%d" % (int(p[0]), int(p[1])) for p in poly)
            parts.extend([
                "  <object>",
                "    <name>%s</name>" % CATEGORY_NAME,
                "    <shape_type>%s</shape_type>"
                % escape(kinds[i] if i < len(kinds) else "polygon"),
                "    <pose>Unspecified</pose><truncated>0</truncated>"
                "<difficult>0</difficult>",
                "    <bndbox><xmin>%d</xmin><ymin>%d</ymin>"
                "<xmax>%d</xmax><ymax>%d</ymax></bndbox>"
                % (int(x0), int(y0), int(x1), int(y1)),
                "    <polygon>%s</polygon>" % pts,
                "  </object>"])
        parts.append("</annotation>")

        stem = os.path.splitext(name)[0] or "image"
        path = os.path.join(out_dir, stem + ".xml")
        ok, err = write_text_atomic(path, "\n".join(parts) + "\n",
                                    keep_backup=False)
        if ok:
            written += 1
        else:
            report.errors.append("%s: %s" % (os.path.basename(path), err))

    report.written.append(out_dir)
    if not written:
        report.warnings.append("no ROI rows to export")
    return report


# ══════════════════════════════════════════════════════════════
# BINARY MASKS
# ══════════════════════════════════════════════════════════════
def export_masks(rows, folder, subdir="export_masks",
                 fill=255, background=0) -> WriteReport:
    """One 8-bit PNG per annotated image, ROIs filled white."""
    report = WriteReport()
    if not HAS_PIL:
        report.errors.append("Pillow is required for mask export")
        return report
    out_dir = os.path.join(str(folder), subdir)
    if not ensure_dir(out_dir):
        report.errors.append("could not create %s" % subdir)
        return report

    written = 0
    for row, polys in _roi_rows(rows):
        width, height = _size_of(row, polys)
        stem = os.path.splitext(row.get("image_name", ""))[0] or "image"
        path = os.path.join(out_dir, stem + "_mask.png")

        def write(tmp, polys=polys, width=width, height=height):
            mask = Image.new("L", (max(1, width), max(1, height)), background)
            drw = ImageDraw.Draw(mask)
            for poly in polys:
                if len(poly) >= 3:
                    drw.polygon([(int(p[0]), int(p[1])) for p in poly], fill=fill)
            mask.save(tmp, format="PNG")

        def verify(tmp):
            with Image.open(tmp) as im:
                im.verify()

        ok, err = atomic_write(path, write, verify, keep_backup=False)
        if ok:
            written += 1
        else:
            report.errors.append("%s: %s" % (os.path.basename(path), err))

    report.written.append(out_dir)
    if not written:
        report.warnings.append("no ROI rows to export")
    return report


# ══════════════════════════════════════════════════════════════
# FILTERING
# ══════════════════════════════════════════════════════════════
def filter_rows(rows, sites=None, cams=None, roi_keys=None, names=None,
                row_types=None):
    """Narrow the row set for a partial export.

    Every argument is optional; a None or empty collection means "no filter
    on this field", so the default is the whole batch."""
    def keep(row):
        if row_types and row.get("row_type") not in row_types:
            return False
        if sites and row.get("site_id") not in sites:
            return False
        if cams and row.get("cam_number") not in cams:
            return False
        if roi_keys and row.get("roi_key") not in roi_keys:
            return False
        if names and row.get("image_name") not in names:
            return False
        return True

    return [r for r in rows if keep(r)]


def distinct(rows, field):
    seen = []
    for row in rows:
        value = str(row.get(field, "") or "")
        if value and value not in seen:
            seen.append(value)
    return sorted(seen, key=geo.natural_key)


# ══════════════════════════════════════════════════════════════
EXPORTERS = {
    "coco": ("COCO instance segmentation (annotations.json)", export_coco),
    "yolo": ("YOLO segmentation labels (one .txt per image)", export_yolo),
    "voc": ("Pascal VOC XML (one file per image)", export_voc),
    "masks": ("Binary mask PNGs (one per image)", export_masks),
}


def run_exports(kinds, rows, folder) -> WriteReport:
    """Run several exporters and merge their reports."""
    combined = WriteReport()
    for kind in kinds:
        entry = EXPORTERS.get(kind)
        if not entry:
            combined.warnings.append("unknown export '%s'" % kind)
            continue
        try:
            combined.merge(entry[1](rows, folder))
        except Exception as exc:                     # pragma: no cover
            combined.errors.append("%s: %s" % (kind, exc))
    return combined

File: packages/roi-core/src/exporters.ts
/**
 * COCO, YOLO, Pascal VOC and binary-mask exports.
 *
 * Port of `apps/roi-studio/roi_studio/core/exporters.py`. The desktop version
 * writes files directly; these functions return the file *contents* and their
 * relative paths, leaving delivery (File System Access, or a ZIP) to the
 * caller. Masks are the one exception: rasterising needs a canvas, so
 * {@link buildMaskSpecs} returns what to draw and the caller draws it.
 *
 * The desktop app hardcodes a single "roi" category. Here each shape carries a
 * class, so the class list drives COCO's `categories`, YOLO's class index and
 * VOC's `<name>`. A batch that uses no classes produces the same single-class
 * output the desktop app produces.
 */

import { APP_VERSION, DEFAULT_CLASS, YOLO_NDIGITS } from "./config.ts";
import * as geo from "./geometry.ts";
import type { MultiPoly, Poly } from "./geometry.ts";
import { pyDumps, pyFloat, type PyValue } from "./pyjson.ts";
import { parseShapeClasses, type CanonRow } from "./store.ts";

export interface ExportFile {
  /** Path relative to the batch folder, e.g. `export_yolo/labels/img1.txt`. */
  path: string;
  text: string;
}

export interface ExportResult {
  files: ExportFile[];
  warnings: string[];
}

export type ExportKind = "coco" | "yolo" | "voc" | "masks";

interface RoiRow {
  row: CanonRow;
  polys: MultiPoly;
}

/** Only rows that actually carry geometry, preferring pixel coordinates. */
function roiRows(rows: readonly CanonRow[]): RoiRow[] {
  const out: RoiRow[] = [];
  for (const row of rows) {
    if (row.row_type !== "roi") continue;
    let polys = geo.parseMultiPolys(row.pixel_coords);
    if (polys.length === 0) {
      const norm = geo.parseMultiPolys(row.normalized_coords);
      const w = Math.trunc(row.image_width || 0);
      const h = Math.trunc(row.image_height || 0);
      if (norm.length && w && h) polys = geo.normToPolys(norm, w, h);
    }
    if (polys.length) out.push({ row, polys });
  }
  return out;
}

/** Image size from the row, falling back to the shapes' own extent. */
function sizeOf(row: CanonRow, polys: MultiPoly): [number, number] {
  const w = Math.trunc(row.image_width || 0);
  const h = Math.trunc(row.image_height || 0);
  if (w && h) return [w, h];
  let maxX = 0;
  let maxY = 0;
  for (const poly of polys) {
    for (const [x, y] of poly) {
      if (x > maxX) maxX = x;
      if (y > maxY) maxY = y;
    }
  }
  return [Math.trunc(maxX) + 1, Math.trunc(maxY) + 1];
}

/** COCO-style [x, y, width, height]. */
function bbox(poly: Poly): [number, number, number, number] {
  const [x0, y0, x1, y1] = geo.polygonBounds(poly);
  return [x0, y0, x1 - x0, y1 - y0];
}

function stemOf(name: string): string {
  const dot = name.lastIndexOf(".");
  const stem = dot > 0 ? name.slice(0, dot) : name;
  return stem || "image";
}

function escapeXml(value: string): string {
  return value
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

/**
 * The class list, in the order that fixes each class's index.
 *
 * The default class is always index 0 when present, so a batch that never used
 * classes exports exactly as the desktop app's single-class output does.
 */
export function classList(rows: readonly CanonRow[]): string[] {
  const seen = new Set<string>();
  for (const { row, polys } of roiRows(rows)) {
    for (const cls of parseShapeClasses(row.shape_classes, polys.length)) {
      seen.add(cls || DEFAULT_CLASS);
    }
  }
  if (seen.size === 0) return [DEFAULT_CLASS];
  const rest = [...seen].filter((c) => c !== DEFAULT_CLASS).sort(geo.naturalCompare);
  return seen.has(DEFAULT_CLASS) ? [DEFAULT_CLASS, ...rest] : rest;
}

function classesFor(row: CanonRow, count: number): string[] {
  return parseShapeClasses(row.shape_classes, count).map((c) => c || DEFAULT_CLASS);
}

// ── COCO ──────────────────────────────────────────────────────

/** One COCO instance-segmentation file for the whole batch. */
export function exportCoco(
  rows: readonly CanonRow[],
  subdir = "export_coco",
  filename = "annotations.json",
  now: Date = new Date(),
): ExportResult {
  const warnings: string[] = [];
  const classes = classList(rows);
  const images: PyValue[] = [];
  const annotations: PyValue[] = [];
  let annId = 1;
  let imageId = 1;

  for (const { row, polys } of roiRows(rows)) {
    const [width, height] = sizeOf(row, polys);
    images.push({
      id: imageId,
      file_name: row.image_name,
      width,
      height,
      roi_key: row.roi_key,
    });
    const kinds = geo.parseShapeTypes(row.shape_types, polys.length);
    const rowClasses = classesFor(row, polys.length);
    polys.forEach((poly, i) => {
      const flat: PyValue[] = [];
      for (const [x, y] of poly) {
        flat.push(pyFloat(x), pyFloat(y));
      }
      const [bx, by, bw, bh] = bbox(poly);
      const cls = rowClasses[i] ?? DEFAULT_CLASS;
      annotations.push({
        id: annId,
        image_id: imageId,
        category_id: classes.indexOf(cls) + 1,
        segmentation: [flat],
        bbox: [pyFloat(bx), pyFloat(by), pyFloat(bw), pyFloat(bh)],
        area: pyFloat(geo.polygonArea(poly)),
        iscrowd: 0,
        shape_type: kinds[i] ?? "polygon",
      });
      annId += 1;
    });
    imageId += 1;
  }

  if (annotations.length === 0) warnings.push("no ROI rows to export");

  const payload: PyValue = {
    info: {
      description: "ROI Studio export",
      version: APP_VERSION,
      date_created: isoSeconds(now),
    },
    licenses: [],
    images,
    annotations,
    categories: classes.map((name, i) => ({
      id: i + 1,
      name,
      supercategory: "region",
    })),
  };

  return {
    files: [{ path: `${subdir}/${filename}`, text: pyDumps(payload, 1) }],
    warnings,
  };
}

function isoSeconds(date: Date): string {
  const pad = (n: number): string => String(n).padStart(2, "0");
  return (
    `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}` +
    `T${pad(date.getHours())}:${pad(date.getMinutes())}:${pad(date.getSeconds())}`
  );
}

// ── YOLO (segmentation format) ────────────────────────────────

/** One .txt of normalised polygons per image, plus a dataset stub. */
export function exportYolo(rows: readonly CanonRow[], subdir = "export_yolo"): ExportResult {
  const warnings: string[] = [];
  const files: ExportFile[] = [];
  const classes = classList(rows);

  for (const { row, polys } of roiRows(rows)) {
    const [width, height] = sizeOf(row, polys);
    const norm = geo.polysToNorm(polys, width, height, YOLO_NDIGITS);
    const rowClasses = classesFor(row, polys.length);
    const lines: string[] = [];
    norm.forEach((poly, i) => {
      if (poly.length < 3) return;
      const coords = poly
        .map(
          ([x, y]) =>
            `${geo.clamp(x, 0, 1).toFixed(6)} ${geo.clamp(y, 0, 1).toFixed(6)}`,
        )
        .join(" ");
      const index = Math.max(0, classes.indexOf(rowClasses[i] ?? DEFAULT_CLASS));
      lines.push(`${index} ${coords}`);
    });
    if (lines.length === 0) continue;
    files.push({
      path: `${subdir}/labels/${stemOf(row.image_name)}.txt`,
      text: `${lines.join("\n")}\n`,
    });
  }

  if (files.length === 0) warnings.push("no ROI rows to export");

  const names = classes.map((name, i) => `  ${i}: ${name}`).join("\n");
  files.push({
    path: `${subdir}/data.yaml`,
    text: `# ROI Studio export\npath: .\ntrain: images\nval: images\n\nnames:\n${names}\n`,
  });

  return { files, warnings };
}

// ── Pascal VOC ────────────────────────────────────────────────

/**
 * One XML per image. VOC describes axis-aligned boxes, so each polygon is
 * written as its bounding box and the exact points are kept in a custom
 * `<polygon>` child for anything that can use them.
 */
export function exportVoc(
  rows: readonly CanonRow[],
  folderName = "",
  subdir = "export_voc",
): ExportResult {
  const warnings: string[] = [];
  const files: ExportFile[] = [];

  for (const { row, polys } of roiRows(rows)) {
    const [width, height] = sizeOf(row, polys);
    const name = row.image_name;
    const parts = [
      "<annotation>",
      `  <folder>${escapeXml(folderName)}</folder>`,
      `  <filename>${escapeXml(name)}</filename>`,
      "  <source><database>ROI Studio</database></source>",
      `  <size><width>${width}</width><height>${height}</height><depth>3</depth></size>`,
      "  <segmented>1</segmented>",
    ];
    const kinds = geo.parseShapeTypes(row.shape_types, polys.length);
    const rowClasses = classesFor(row, polys.length);
    polys.forEach((poly, i) => {
      const [x0, y0, x1, y1] = geo.polygonBounds(poly);
      const pts = poly.map(([x, y]) => `${Math.trunc(x)},${Math.trunc(y)}`).join(" ");
      parts.push(
        "  <object>",
        `    <name>${escapeXml(rowClasses[i] ?? DEFAULT_CLASS)}</name>`,
        `    <shape_type>${escapeXml(kinds[i] ?? "polygon")}</shape_type>`,
        "    <pose>Unspecified</pose><truncated>0</truncated><difficult>0</difficult>",
        `    <bndbox><xmin>${Math.trunc(x0)}</xmin><ymin>${Math.trunc(y0)}</ymin>` +
          `<xmax>${Math.trunc(x1)}</xmax><ymax>${Math.trunc(y1)}</ymax></bndbox>`,
        `    <polygon>${pts}</polygon>`,
        "  </object>",
      );
    });
    parts.push("</annotation>");
    files.push({ path: `${subdir}/${stemOf(name)}.xml`, text: `${parts.join("\n")}\n` });
  }

  if (files.length === 0) warnings.push("no ROI rows to export");
  return { files, warnings };
}

// ── binary masks ──────────────────────────────────────────────

export interface MaskSpec {
  /** Path relative to the batch folder, e.g. `export_masks/img1_mask.png`. */
  path: string;
  width: number;
  height: number;
  polys: MultiPoly;
}

/**
 * What to rasterise for each annotated image: one 8-bit PNG with the ROIs
 * filled. Drawing needs a canvas, so the caller does it — see
 * `apps/frontend/services/polygonAnnotation.service.ts`.
 */
export function buildMaskSpecs(
  rows: readonly CanonRow[],
  subdir = "export_masks",
): { specs: MaskSpec[]; warnings: string[] } {
  const specs: MaskSpec[] = [];
  for (const { row, polys } of roiRows(rows)) {
    const [width, height] = sizeOf(row, polys);
    const usable = polys.filter((poly) => poly.length >= 3);
    if (usable.length === 0) continue;
    specs.push({
      path: `${subdir}/${stemOf(row.image_name)}_mask.png`,
      width: Math.max(1, width),
      height: Math.max(1, height),
      polys: usable,
    });
  }
  const warnings = specs.length === 0 ? ["no ROI rows to export"] : [];
  return { specs, warnings };
}

// ── filtering helpers ─────────────────────────────────────────

export interface RowFilter {
  sites?: readonly string[];
  cams?: readonly string[];
  roiKeys?: readonly string[];
  names?: readonly string[];
  rowTypes?: readonly string[];
}

/** Narrow an export to particular sites, cameras, keys, names or row types. */
export function filterRows(
  rows: readonly CanonRow[],
  filter: RowFilter = {},
): CanonRow[] {
  const { sites, cams, roiKeys, names, rowTypes } = filter;
  return rows.filter((row) => {
    if (sites?.length && !sites.includes(row.site_id)) return false;
    if (cams?.length && !cams.includes(row.cam_number)) return false;
    if (roiKeys?.length && !roiKeys.includes(row.roi_key)) return false;
    if (names?.length && !names.includes(row.image_name)) return false;
    if (rowTypes?.length && !rowTypes.includes(row.row_type)) return false;
    return true;
  });
}

/** Unique non-empty values of one field, in natural order. */
export function distinct(rows: readonly CanonRow[], field: keyof CanonRow): string[] {
  const seen = new Set<string>();
  for (const row of rows) {
    const value = String(row[field] ?? "").trim();
    if (value) seen.add(value);
  }
  return [...seen].sort(geo.naturalCompare);
}

/** Run a subset of the exporters and merge their results. */
export function runExports(
  kinds: readonly ExportKind[],
  rows: readonly CanonRow[],
  folderName = "",
  now: Date = new Date(),
): ExportResult & { masks: MaskSpec[] } {
  const files: ExportFile[] = [];
  const warnings: string[] = [];
  const masks: MaskSpec[] = [];
  for (const kind of kinds) {
    if (kind === "coco") {
      const result = exportCoco(rows, "export_coco", "annotations.json", now);
      files.push(...result.files);
      warnings.push(...result.warnings);
    } else if (kind === "yolo") {
      const result = exportYolo(rows);
      files.push(...result.files);
      warnings.push(...result.warnings);
    } else if (kind === "voc") {
      const result = exportVoc(rows, folderName);
      files.push(...result.files);
      warnings.push(...result.warnings);
    } else if (kind === "masks") {
      const result = buildMaskSpecs(rows);
      masks.push(...result.specs);
      warnings.push(...result.warnings);
    }
  }
  return { files, warnings, masks };
}
